### DISCUSSION TITLE
Consider adding channel mechanics from Golang

### DISCUSSION THREAD

**Discussion by Banyc** [dotnet/csharplang#5038] [Language Ideas]
## Reasons

With channel mechanics, async code could be written in serial manner. It could also further boost the participation of C# in async scenarios.

The "select channels" way seems not easy to be implemented under the current C# syntax. I have found a random snippet in [a website](https://csharp.hotexamples.com/examples/-/WaitGroup/-/php-waitgroup-class-examples.html):

```csharp
await Go.Select(
    Go.Case(results, r => {
        Console.WriteLine($"Got {r.Value} for {r.Path}");
        totalFiles++;
    }),
    Go.Case(errors, exception => {
        Console.WriteLine($"EXCEPTION: {exception}");
        totalFiles++;
    }));
```

**Comment by HaloFour** [comment]
See: https://github.com/dotnet/roslyn/issues/5187
